### PR TITLE
Build with libcxx(abi) 15

### PR DESCRIPTION
**Pull request by rschu1ze**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

This is a prerequisite for #42245

Multiple things happen in this PR:
- Submodules libcxx and libcxxabi are updated from 14 to 15, see https://github.com/ClickHouse/libcxx/pull/14 and https://github.com/ClickHouse/libcxxabi/pull/5
- This broke some stuff:
   - in submodule rocksdb: fixed with https://github.com/ClickHouse/rocksdb/pull/17
   - in submodule vectorscan: fixed by patching stuff in libcxx, one of the two problems was reported to upstream here: https://github.com/VectorCamp/vectorscan/issues/134
   - in ClickHouse, the main problem was that with libcxx 15 the compiler complains if typedefs are created for incomplete types. I am not sure if the standard mandates such behavior and if it previously worked by chance. Anyways, this problem hit in the aggregation code and in clickhouse-copier. I untangled the code so that it compiles now (as a side effect, compilation should now also be a little bit faster). These were mere refactorings, there are no semantic changes to the code.

### CODE REVIEW

File: src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.h [hunk 1]
```diff
@@ -23,8 +25,69 @@ class AggregatingSortedAlgorithm final : public IMergingAlgorithmWithDelayedChun
     void consume(Input & input, size_t source_num) override;
     Status merge() override;
 
-    struct SimpleAggregateDescription;
-    struct AggregateDescription;
```
Comment (antonio2368) [inline src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.h:27]: How did it work before 😮?

Comment (rschu1ze) [reply src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.h:27]: I don't fully understand it by myself but I think it has to do with slightly different inlining with libcxx 15.

The problem is even worse with clickhouse-copier ... it fell into thousands of pieces and I try to fix it.

**Comment by rschu1ze** [comment]
`clickhouse-client --query "select 1"` coredumps in fasttest. Hm, works nicely for me locally.

**Comment by rschu1ze** [comment]
server.log:
```
2022.10.20 22:30:52.537137 [ 20411 ] {} <Fatal> BaseDaemon: Stack trace: 0xfa0a22f 0xfa03ef6 0xf972394 0xf97275f 0xb9582d6 0xb957926 0x10de455b 0x10de3981 0xb94706f 0x11043186 0xb93c2f0 0x110571b2 0xb9399a0 0x6fd47f1 0x7ff1ce4c2083 0x6fd3cae
2022.10.20 22:30:52.562551 [ 20411 ] {} <Fatal> BaseDaemon: 3.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/atomic:958: bool std::__1::__cxx_atomic_load[abi:v15003]<bool>(std::__1::__cxx_atomic_base_impl<bool> const*, std::__1::memory_order)
2022.10.20 22:30:52.562598 [ 20411 ] {} <Fatal> BaseDaemon: 3.2. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/atomic:1560: std::__1::__atomic_base<bool, false>::load[abi:v15003](std::__1::memory_order) const
2022.10.20 22:30:52.562602 [ 20411 ] {} <Fatal> BaseDaemon: 3.3. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/atomic:1564: std::__1::__atomic_base<bool, false>::operator bool[abi:v15003]() const
2022.10.20 22:30:52.562607 [ 20411 ] {} <Fatal> BaseDaemon: 3.4. inlined from /fasttest-workspace/build/../../ClickHouse/src/Common/ThreadPool.h:228: ~ThreadFromGlobalPoolImpl
2022.10.20 22:30:52.562610 [ 20411 ] {} <Fatal> BaseDaemon: 3. /fasttest-workspace/build/../../ClickHouse/src/Interpreters/ExternalLoader.cpp:1235: DB::ExternalLoader::PeriodicUpdater::enable(bool) @ 0xfa0a22f in /fasttest-workspace/build/programs/clickhouse
2022.10.20 22:30:52.569182 [ 20411 ] {} <Fatal> BaseDaemon: 4.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/__memory/shared_ptr.h:1508: weak_ptr<const DB::Context>
2022.10.20 22:30:52.569207 [ 20411 ] {} <Fatal> BaseDaemon: 4. /fasttest-workspace/build/../../ClickHouse/src/Interpreters/ExternalDictionariesLoader.cpp:26: DB::ExternalDictionariesLoader::ExternalDictionariesLoader(std::__1::shared_ptr<DB::Context const>) @ 0xfa03ef6 in /fasttest-workspace/build/programs/clickhouse
2022.10.20 22:30:52.641362 [ 20411 ] {} <Fatal> BaseDaemon: 5. /fasttest-workspace/build/../../ClickHouse/src/Interpreters/Context.cpp:1552: DB::Context::getExternalDictionariesLoaderUnlocked() @ 0xf972394 in /fasttest-workspace/build/programs/clickhouse
2022.10.20 22:30:52.711380 [ 20411 ] {} <Fatal> BaseDaemon: 6. /fasttest-workspace/build/../../ClickHouse/src/Interpreters/Context.cpp:0: DB::Context::loadOrReloadDictionaries(Poco::Util::AbstractConfiguration const&) @ 0xf97275f in /fasttest-workspace/build/programs/clickhouse
2022.10.20 22:30:52.756343 [ 20411 ] {} <Fatal> BaseDaemon: 7.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/poco/Foundation/include/Poco/AutoPtr.h:231: Poco::AutoPtr<Poco::Util::AbstractConfiguration>::operator*()
2022.10.20 22:30:52.756377 [ 20411 ] {} <Fatal> BaseDaemon: 7. /fasttest-workspace/build/../../ClickHouse/programs/server/Server.cpp:1170: DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_4::operator()(Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool) const @ 0xb9582d6 in /fasttest-workspace/build/programs/clickhouse
2022.10.20 22:30:52.801080 [ 20411 ] {} <Fatal> BaseDaemon: 8.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/poco/Foundation/include/Poco/AutoPtr.h:90: ~AutoPtr
2022.10.20 22:30:52.801119 [ 20411 ] {} <Fatal> BaseDaemon: 8.2. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/__functional/invoke.h:394: decltype(std::declval<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_4&>()(std::declval<Poco::AutoPtr<Poco::Util::AbstractConfiguration>>(), std::declval<bool>())) std::__1::__invoke[abi:v15003]<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_4&, Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool>(DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_4&, Poco::AutoPtr<Poco::Util::AbstractConfiguration>&&, bool&&)
2022.10.20 22:30:52.801146 [ 20411 ] {} <Fatal> BaseDaemon: 8.3. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/__functional/invoke.h:479: void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_4&, Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool>(DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_4&, Poco::AutoPtr<Poco::Util::AbstractConfiguration>&&, bool&&)
2022.10.20 22:30:52.801157 [ 20411 ] {} <Fatal> BaseDaemon: 8.4. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/__functional/function.h:235: std::__1::__function::__default_alloc_func<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_4, void (Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool)>::operator()[abi:v15003](Poco::AutoPtr<Poco::Util::AbstractConfiguration>&&, bool&&)
2022.10.20 22:30:52.801160 [ 20411 ] {} <Fatal> BaseDaemon: 8. /fasttest-workspace/build/../../ClickHouse/contrib/libcxx/include/__functional/function.h:716: void std::__1::__function::__policy_invoker<void (Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool)>::__call_impl<std::__1::__function::__default_alloc_func<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_4, void (Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool)>>(std::__1::__function::__policy_storage const*, Poco::AutoPtr<Poco::Util::AbstractConfiguration>&&, bool) @ 0xb957926 in /fasttest-workspace/build/programs/clickhouse
2022.10.20 22:30:52.809578 [ 20411 ] {} <Fatal> BaseDaemon: 9.1. inlined from /fasttest-workspace/build/../../ClickHouse/contrib/poco/Foundation/include/Poco/AutoPtr.h:90: ~AutoPtr
```

I have seen this error also in other PRs.

File: src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.h (hunk 1 above)
Comment (KochetovNicolai) [reply src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.h:27]: I think it was possible to declare `std::vector<AggregateDescription>` cause you did not need to know `sizeof(AggregateDescription)` in order to calculate `sizeof(std::vector<AggregateDescription>)`

File: src/Processors/Merges/Algorithms/SummingSortedAlgorithm.cpp [hunk 2]
```diff
@@ -23,10 +23,6 @@ namespace ErrorCodes
     extern const int CORRUPTED_DATA;
 }
 
-SummingSortedAlgorithm::ColumnsDefinition::ColumnsDefinition() = default;
```
Comment (KochetovNicolai) [inline src/Processors/Merges/Algorithms/SummingSortedAlgorithm.cpp:26]: :(

File: src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.h (hunk 1 above)
Comment (antonio2368) [reply src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.h:27]: that's the interesting part for me, why do they need size now 🤔

File: src/Processors/Merges/Algorithms/SummingSortedAlgorithm.cpp (hunk 2 above)
Comment (rschu1ze) [reply src/Processors/Merges/Algorithms/SummingSortedAlgorithm.cpp:26]: `ColumnsDefinition` contains a `std::vector<AggregateDescription>` member. The vector destructor `std::~vector()` (called in `ColumnsDefinition::~ColumnsDefinition()`) does some bookkeeping which requires to know the full definition of `AggregateDescription`. To make this work, the three defaulted functions are here moved behind the definition of `AggregateDescription'. So I think the change makes sense in general.

The question remains why it used to work before. I see that they made a lot of methods of `std::vector` in libcxx15 `constexpr`, including the vector destructor. However, that would only be an explanation if `ColumnsDefinition` is used in a compile-time/constexpr context but this is not done here. Besides that and nothing else changed in `std::vector`, I don't really have a good explanation :-/

**Comment by rschu1ze** [comment]
I think this PR is ready-to-merge. Does anyone like to approve? Thanks!

**Review by devcrafter — APPROVED** [review]
LGTM

File: src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.h (hunk 1 above)
Comment (azat) [reply src/Processors/Merges/Algorithms/AggregatingSortedAlgorithm.h:27]: Looks like it had been introduced here - https://github.com/llvm/llvm-project/commit/f4fb72e6d4cee1097e6606e66232fe55e793cd86

**Pull request merged**